File: Codeforce/Div2/Round 936/C.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(200000)


def dfs(a, par, size):
    global comp
    curdeg = 1
    for g in links[a]:
        if g == par:
            continue
        result = dfs(g, a, size)
        if result >= size:
            comp += 1
        else:
            curdeg += result
    return curdeg


for TC in range(int(input())):
    N, K = map(int, input().split())
    links = [[] for i in range(N)]
    for i in range(N-1):
        s, e = map(int, input().split())
        links[s-1].append(e-1)
        links[e-1].append(s-1)

    left, right = 1, N
    ans = 1
    while (left <= right):
        mid = (left+right)//2

        comp = 0
        result = dfs(0, -1, mid)
        if result >= mid:
            comp += 1
        if comp > K:
            ans = mid
            left = mid+1
        else:
            right = mid-1
    print(ans)

File: Codeforce/Div2/Round 936/C.cpp
#include <bits/stdc++.h>
using namespace std;

void fastio(){
    cin.tie(0);
    cout.tie(0);
    ios_base::sync_with_stdio(false);
}

int comp;
vector<vector<int>> links;

int dfs(int a, int par, int size) {
    int curdeg = 1;
    for (int g : links[a]) {
        if (g == par) continue;
        int result = dfs(g, a, size);
        if (result >= size) {
            comp++;
        } else {
            curdeg += result;
        }
    }
    return curdeg;
}


int main(){
    fastio();int T;
    cin >> T;
    while (T--) {
        int N, K;
        cin >> N >> K;
        links.assign(N, vector<int>());

        for (int i = 0; i < N - 1; ++i) {
            int s, e;
            cin >> s >> e;
            --s; --e;
            links[s].emplace_back(e);
            links[e].emplace_back(s);
        }

        int left = 1, right = N, ans = 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            comp = 0;
            int result = dfs(0, -1, mid);
            if (result >= mid) {
                comp++;
            }
            if (comp > K) {
                ans = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        cout << ans << '\n';
    }
    return 0;
}